File: Python/ExampleTests/common/catalog_item_string.py
from . import tokens
DNC_STRING = '?DNC?'


class CatalogItemString:
    def __init__(self, name: str = '', price: str = ''):
        self.name = name
        self.price = price

    @classmethod
    def from_list(cls, values):
        v = list(values)
        return cls(
            v[0] if len(v) > 0 else '',
            v[1] if len(v) > 1 else ''
        )

    @classmethod
    def from_text(cls, text):
        """Builds from the text form, e.g. Money as '25 USD'."""
        parts = tokens.require(text, 2, 'CatalogItem')
        return cls(
            parts[0],
            parts[1]
        )

    def __str__(self):
        return (tokens.token(self.name) + ' ' +
                tokens.token(self.price))

    def _key(self):
        return (self.name, self.price)

    def __eq__(self, other):
        if not isinstance(other, CatalogItemString):
            return NotImplemented
        return all(a == b or a == DNC_STRING or b == DNC_STRING
                   for a, b in zip(self._key(), other._key()))

    def __hash__(self):
        return hash(self._key())

    def __repr__(self):
        return f'CatalogItemString({self})'

File: Swift/ExampleTests/common/CatalogItemString.swift
public struct CatalogItemString: CustomStringConvertible, Equatable {
    public let name: String
    public let price: String

    public init(name: String, price: String) {
        self.name = name
        self.price = price
    }

    public init(fromArray v: [String]) {
        self.name = v.count > 0 ? v[0] : ""
        self.price = v.count > 1 ? v[1] : ""
    }

    /// Builds from the text form, e.g. Money as "25 USD".
    public static func fromText(_ text: String) -> CatalogItemString {
        let parts = Tokens.require(text, 2, "CatalogItem")
        return CatalogItemString(name: parts[0], price: parts[1])
    }

    public var description: String {
        return Tokens.token(name) + " " + Tokens.token(price)
    }

    public static let dncString = "?DNC?"

    public static func == (a: CatalogItemString, b: CatalogItemString) -> Bool {
        return (a.name == dncString || b.name == dncString || a.name == b.name)
            && (a.price == dncString || b.price == dncString || a.price == b.price)
    }
}
